Kernel/VirtIO/VirtIOConsole.cpp

```diff
@@ -73,115 +60,135 @@ bool VirtIOConsole::handle_device_config_change()
 
 void VirtIOConsole::handle_queue_update(u16 queue_index)
 {
-    dbgln_if(VIRTIO_DEBUG, "VirtIOConsole: Handle queue update");
-    VERIFY(queue_index <= TRANSMITQ);
-    switch (queue_index) {
-    case RECEIVEQ: {
-        ScopedSpinLock ringbuffer_lock(m_transmit_buffer->lock());
-        auto& queue = get_queue(RECEIVEQ);
+    dbgln_if(VIRTIO_DEBUG, "VirtIOConsole: Handle queue update {}", queue_index);
+
+    if (queue_index == CONTROL_RECEIVEQ) {
+        ScopedSpinLock ringbuffer_lock(m_control_receive_buffer->lock());
+        auto& queue = get_queue(CONTROL_RECEIVEQ);
         ScopedSpinLock queue_lock(queue.lock());
         size_t used;
         VirtIOQueueChain popped_chain = queue.pop_used_buffer_chain(used);
 
-        auto used_space = m_receive_buffer->reserve_space(used).value();
-        auto remaining_space = RINGBUFFER_SIZE - used;
-
-        // Our algorithm always has only one buffer in the queue.
-        size_t unused;
-        VERIFY(queue.pop_used_buffer_chain(unused).is_empty());
-        popped_chain.release_buffer_slots_to_queue();
+        while (!popped_chain.is_empty()) {
+            popped_chain.for_each([&](auto addr, auto) {
+                auto offset = addr.as_ptr() - m_control_receive_buffer->start_of_region().as_ptr();
+                ControlMessage* message = (ControlMessage*)m_control_receive_buffer->vaddr().offset(offset).as_ptr();
+                process_control_message(*message);
+            });
 
-        VirtIOQueueChain new_chain(queue);
-        if (remaining_space != 0) {
-            new_chain.add_buffer_to_chain(used_space.offset(used), remaining_space, BufferType::DeviceWritable);
+            supply_chain_and_notify(CONTROL_RECEIVEQ, popped_chain);
+            popped_chain = queue.pop_used_buffer_chain(used);
         }
-        supply_chain_and_notify(RECEIVEQ, new_chain);
-
-        evaluate_block_conditions();
-        break;
-    }
-    case TRANSMITQ: {
-        ScopedSpinLock ringbuffer_lock(m_transmit_buffer->lock());
-        auto& queue = get_queue(TRANSMITQ);
+    } else if (queue_index == CONTROL_TRANSMITQ) {
+        ScopedSpinLock ringbuffer_lock(m_control_transmit_buffer->lock());
+        auto& queue = get_queue(CONTROL_TRANSMITQ);
         ScopedSpinLock queue_lock(queue.lock());
         size_t used;
         VirtIOQueueChain popped_chain = queue.pop_used_buffer_chain(used);
```

Comment:
this should be moved into the do-while loop and the last line of the do-while loop should be removed to avoid duplication

Reply:
Calling release_buffer_sots_to_queue on a queuechain will empty the chain, so you might end up missing chains. You need to test the popped chain immediately after popping the chain (but before you empty it) to determine if the chain had content.

Reply:
we already call `for_each` before the release call, so that takes care of it.